Scan build warnings 2

ryanhsu19 opened the pull request:
Third-party files have been removed from the list. See #2379 for details.

jhand2 commented on samples/file-encryptor/enclave/keys.cpp:
```diff
@@ -290,9 +290,17 @@ int ecall_dispatcher::prepare_encryption_header(
         TRACE_ENCLAVE("password_key");
         for (unsigned int i = 0; i < ENCRYPTION_KEY_SIZE_IN_BYTES; i++)
+        {
+            if (password_key[i] > UINT32_MAX)
```
Discussed previously, but it is not possible for this condition to be true. What warning is this addressing?

jhand2 commented on samples/file-encryptor/enclave/keys.cpp:
```diff
@@ -290,9 +290,17 @@ int ecall_dispatcher::prepare_encryption_header(
             TRACE_ENCLAVE(
                 "password_key[%d] =0x%02x", i, (unsigned int)(password_key[i]));
-        goto exit;
+            goto exit;
```
I think this goto should be outsize the loop

jhand2 commented on tests/crypto/asn1_tests.c:
```diff
@@ -264,8 +264,6 @@ static oe_result_t _parse(oe_asn1_t* asn1)
                 if (result != OE_OK)
                     goto done;
 
-                result = _parse(&sequence);
```
I think this line is supposed to be kept, it is trying to recursively parse sequence.

The missing thing here is that the response isn't used. So the whole case should be:
```
case OE_ASN1_TAG_SEQUENCE:
{
    oe_asn1_t sequence;

    result = oe_asn1_get_sequence(asn1, &sequence);
    if (result != OE_OK)
        goto done;

    result = _parse(&sequence);
    if (result != OE_OK)
        goto done;

    break;
}
```

jhand2 commented on tests/libunwind/enc/enc.c:
```diff
@@ -65,10 +65,6 @@ int test(char test_name[201], uint32_t pid)
-    if (!(__environ = (char**)calloc(1, sizeof(char**))))
-    {
-        rval = 1;
-    }
```
This check should be kept. Just change `rval = 1;` to `return 1;`

ryanhsu19 commented:
Thank you! I will keep working on other bugs.

radhikaj commented:
bors try

jhand2 commented on enclave/core/sgx/td.c:
```diff
@@ -91,19 +91,23 @@ void td_push_callsite(td_t* td, Callsite* callsite)
-        // The outermost ecall is about to return.
-        // Clear the thread-local storage.
-        td_clear(td);
+        oe_abort();
```
scan-build doesn't know that `oe_abort` stops execution. So it thinks td is potentially null and that we just ignore the result of the check.

mingweishih replied:
I see. This is unfortunate as I like the original pattern better :)

jhand2 commented:
Hey @ryanhsu19 can you clean up the commit history here? We shouldn't have the merge commit or the extra "check for unsigned int added by Ryan Hsu 12/6/2019" commit. The other two commits should just be squashed into one.

Let me know if you need help with how to do this.

jhand2 commented on enclave/core/sgx/report.c:
```diff
@@ -210,6 +210,9 @@ oe_result_t oe_get_remote_report(
     sgx_quote = (sgx_quote_t*)report_buffer;
 
+    if (sgx_quote == NULL)
+        OE_RAISE(OE_UNEXPECTED);
```
We should raise `OE_INVALID_PARAMETER` instead of unexpected

jhand2 commented on enclave/sgx/revocationinfo.c:
```diff
@@ -81,20 +82,30 @@ oe_result_t oe_get_revocation_info(oe_get_revocation_info_args_t* args)
+            else
+            {
+                in.tcb_info = temp;
+            }
```
What is the reason for doing this? There is no memory safety issue in the original implementation that i can see.

jhand2 commented on host/sgx/elf.c:
```diff
@@ -1674,7 +1677,10 @@ int elf64_add_section(
         elf64_ehdr_t* ehdr = _get_header(elf);
         if (oe_safe_add_u16(ehdr->e_shnum, 1, &ehdr->e_shnum) != OE_OK)
+        {
+            mem_free(&mem);
```
This should just be done once after the `done` label

jhand2 commented on tests/libc/enc/enc.c:
```diff
@@ -111,7 +111,7 @@ int run_test(const char* name, int (*main)(int argc, const char* argv[]))
     /* Allocate an environment for invoking the test. */
     {
-        if (!(environ = (char**)calloc(1, sizeof(char**))))
+        if (!(environ = (char**)calloc(1, sizeof(char*))))
```
why change this? They're functionally equivalent and its general practice to use the sizeof the type you're allocating

jhand2 replied:
This is not resolved

ryanhsu19 replied:
I know char* and char** are both pointers, but the meanings are different, and it does generate 10 related sizeof operand mismatch. Therefore I am wondering why char** is designed to point to an array of one char** element.

jhand2 replied:
oh, sorry, you're right. Re-resolving :smile:

jhand2 commented on tests/libunwind/enc/enc.c:
```diff
@@ -65,9 +65,9 @@ int test(char test_name[201], uint32_t pid)
     g_pid = pid;
     printf("RUNNING: %s\n", __TEST__);
-    if (!(__environ = (char**)calloc(1, sizeof(char**))))
+    if (!(__environ = (char**)calloc(1, sizeof(char*))))
```
same as above

jhand2 replied:
Also not resolved

jhand2 replied:
My mistake. Resolving

jhand2 commented on tests/memory/enc/basic.c:
```diff
@@ -38,6 +38,8 @@ void test_malloc(void)
     ptr = (int*)malloc(~((size_t)0));
     OE_TEST(ptr == NULL);
+    if (ptr)
+        free(ptr);
```
This code is unreachable. Do the check/free before OE_TEST

jhand2 commented on tests/memory/enc/basic.c:
```diff
@@ -102,11 +104,8 @@ void test_realloc(void)
-    ptr = realloc(NULL, 0);
-    free(ptr);
+    if (ptr2)
+        free(ptr2);
```
Same as above

jhand2 commented on tests/stdc/enc/enc.cpp:
```diff
@@ -274,6 +274,10 @@ int test(char buf1[BUFSIZE], char buf2[BUFSIZE])
     OE_TEST(p == NULL);
+
+    if (p)
+        free(p);
```
unreachable

jhand2 commented:
It would appear you accidentally pulled in some extra commits that pull a bunch of stuff into this PR

ryanhsu19 commented:
Hi Jordan,

I will remove the last three ones for now. I changed it because it showed a scan-build error regarding the OE_TEST conditions, but again, OE_TEST runs oe_abort() once the condition is false.

As for the calloc, there is a type-mismatching error. What Alvin and I were thinking was, since char** is an array of char*, we discussed and changed sizeof(char**) to sizeof(char*), and the scan-build error has disappeared since then.

The revocation info part is for the memory leak issues.

I will change them back.

Ryan

From: Jordan <notifications@github.com>
Sent: Monday, January 13, 2020 11:44 AM
To: openenclave/openenclave <openenclave@noreply.github.com>
Cc: Ryan Hsu <Ryan.Hsu@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [openenclave/openenclave] Scan build warnings 2 (#2407)


@jhand2 requested changes on this pull request.

A few comments

________________________________

In enclave/core/sgx/report.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365981456&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373634883&sdata=3CUw6WyeVBuvEHDBWdPlNpK%2FfFFR1VkBHYjgtSkQo%2Bc%3D&reserved=0>:

> +    if (sgx_quote == NULL)

+        OE_RAISE(OE_UNEXPECTED);

We should raise OE_INVALID_PARAMETER instead of unexpected

________________________________

In enclave/sgx/revocationinfo.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365982628&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373644836&sdata=nytbDKlZ5h%2BlKBXbW2npVyP4US9n9VqYltMIafaHbq8%3D&reserved=0>:

> +            if (!(temp = realloc(in.tcb_info, out.tcb_info_size)))

+            {

                 OE_RAISE(OE_OUT_OF_MEMORY);

+            }

+            else

+            {

+                in.tcb_info = temp;

+            }

What is the reason for doing this? There is no memory safety issue in the original implementation that i can see.

________________________________

In host/sgx/elf.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365985176&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373644836&sdata=vsk46JQyXL2Uue0PAxhThRmg%2BLfHMhPXnL15LilbS3Q%3D&reserved=0>:

> @@ -1674,7 +1677,10 @@ int elf64_add_section(

         /* Update number of sections */

         elf64_ehdr_t* ehdr = _get_header(elf);

         if (oe_safe_add_u16(ehdr->e_shnum, 1, &ehdr->e_shnum) != OE_OK)

+        {

+            mem_free(&mem);

This should just be done once after the done label

________________________________

In tests/libc/enc/enc.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365989232&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373654788&sdata=XikqaXXQ%2FovTj6Ky8Z99hkHoFd7xf6GMxKBqH9C%2BkGI%3D&reserved=0>:

> @@ -111,7 +111,7 @@ int run_test(const char* name, int (*main)(int argc, const char* argv[]))



     /* Allocate an environment for invoking the test. */

     {

-        if (!(environ = (char**)calloc(1, sizeof(char**))))

+        if (!(environ = (char**)calloc(1, sizeof(char*))))

why change this? They're functionally equivalent and its general practice to use the sizeof the type you're allocating

________________________________

In tests/libunwind/enc/enc.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365989332&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373654788&sdata=W1uY0FrIsbW59FoJXHpkr0YeOdS3yIySMIBRC8i3bwI%3D&reserved=0>:

> @@ -65,9 +65,9 @@ int test(char test_name[201], uint32_t pid)

     int rval = 1;

     g_pid = pid;

     printf("RUNNING: %s\n", __TEST__);

-    if (!(__environ = (char**)calloc(1, sizeof(char**))))

+    if (!(__environ = (char**)calloc(1, sizeof(char*))))

same as above

________________________________

In tests/memory/enc/basic.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365989528&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373664746&sdata=zyO6ImGmGj0Az4k6fxeuFf25gtxFQcE5Df86YIykmto%3D&reserved=0>:

> +    if (ptr)

+        free(ptr);

This code is unreachable. Do the check/free before OE_TEST

________________________________

In tests/memory/enc/basic.c<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365989628&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373664746&sdata=0T7xxubhSPVxQUhL1Tu%2BjBIE0%2BerNqODeNyCo3OEHKg%3D&reserved=0>:

> @@ -102,11 +104,8 @@ void test_realloc(void)

 #pragma GCC diagnostic pop

 #endif

     OE_TEST(ptr2 == NULL);

-

-    /* realloc(X, 0) is implementation defined, but we can always free it. */

-    free(ptr);

-    ptr = realloc(NULL, 0);

-    free(ptr);

+    if (ptr2)

+        free(ptr2);

Same as above

________________________________

In tests/stdc/enc/enc.cpp<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%23discussion_r365991606&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373664746&sdata=5fvxZ2NZwCpjqzy8p0Oyr9rExntjJ7gxQQjfvSaQ4Xs%3D&reserved=0>:

> +    if (p)

+        free(p);

unreachable

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenenclave%2Fopenenclave%2Fpull%2F2407%3Femail_source%3Dnotifications%26email_token%3DAN2IHTCRQRK3VW3JEN2ZC6DQ5S77XA5CNFSM4KDMDTEKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCRR3GYQ%23pullrequestreview-342078306&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373674700&sdata=AtWUQIY8Y1fIh7e06hMh%2F9GoIpquPZ%2F%2FTkzjbT0RowU%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAN2IHTHG6AQLI23BB22XIZDQ5S77XANCNFSM4KDMDTEA&data=02%7C01%7CRyan.Hsu%40microsoft.com%7Ce1b1d3a87778447e057f08d79860ed22%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637145414373674700&sdata=wvZ%2FZYIGLiaW8sh61cNFtobGwgXjxkGz%2F7V8IST0Qfc%3D&reserved=0>.

dthaler commented on common/sgx/verifier.c:
```diff
@@ -39,8 +39,10 @@ static oe_result_t _on_unregister(oe_attestation_role_t* context)
 {
-    oe_free(claim->name);
-    oe_free(claim->value);
+    if (claim->name != NULL)
```
The standard free() is specified to be a no-op if NULL is passed, and so should oe_free.
Hence I think this change should not be needed and it should be annotated however free() is annotated so that no warnings are given when NULL is passed.

jhand2 replied:
Indeed both `oe_free` and `free` check for null ptr.

@ryanhsu19 can you share the warning generated for these?

ryanhsu19 replied:
I included it because previously scan-build popped up some double-free warnings for many files, but I could not reconstruct it so I will remove it.

jhand2 replied:
Fixing double free would require setting `claim->name` and `claim->value` to NULL after free. Not checking for NULL.

anakrish replied:
Couple of questions:

1. Why does scanbuild not care that the `claim` variable is not null-checked before dereferencing (`claim-<name`)?

2. Why does scanbuild care that the parameter to oe_free be null checked. It is valid to pass NULL value to functions.

jhand2 replied:
1. Scan build only catches bugs that can actually happen in the execution path. So if no code ever passes null to `_free_claim` then scan-build won't warn about it.
2. It doesn't. I don't think this fixes the reported warning.

dthaler commented on enclave/crypto/cert.c:
```diff
@@ -764,8 +764,8 @@ oe_result_t oe_cert_chain_read_pem(
-
-    _referent_free(referent);
+    if (referent)
+        _referent_free(referent);
```
Same comment here about _referent_free(), it should a no-op if NULL is passed so this change should not be needed.  Fix the header so no warning is generated.

ryanhsu19 replied:
If you enter from mbedtls_x509_crt_free(referent->crt) and referent is a null pointer, then access to field crt in _referent_free() results in a dereference of a null pointer, therefore the warning.

jhand2 replied:
_referent_free does not check for a null pointer. We should fix the implementation to check there.

ryanhsu19 replied:
Okay. Working on it now.

dthaler replied:
+1 to fixing _referent_free to check for a null pointer, for consistent with the rest of the *free() functions

dthaler commented on enclave/sgx/revocationinfo.c:
```diff
@@ -160,10 +158,12 @@ void oe_free_get_revocation_info_args(oe_get_revocation_info_args_t* args)
         for (size_t i = 0; i < OE_COUNTOF(args->crl); i++)
-            free(args->crl[i]);
+            if (args->crl[i])
+                free(args->crl[i]);
```
This was already safe per specification of standard free().  Do we know why it generated a warning?  Rather than changing all code that is already safe, it's better to have a fix that stops generating bad warnings.

ryanhsu19 replied:
I included it because scan-build popped up many double-free warnings. The main purpose for the file edit is to add in-buffers into the free buffer section in lines 140-149 to prevent potential memory leaks.

jhand2 replied:
So if there are double free warnings, does this really solve them? `free` does not set the pointer to null, so even with this check the same double-free bug would persist.

ryanhsu19 commented:
As for enclave/sgx/revocationinfo.c, I made some changes due to previously many double-free warnings (there is only one such warning now and it is for a third-party file), and I included the change this file as I added the in-buffers into the /* free buffers */ section in lines 140-149. Same for common/sgx/verifier.c.

As for the enclave/crypto/cert.c, if you enter from mbedtls_x509_crt_free(referent->crt) and referent is a null pointer, then access to field crt in _referent_free() results in a dereference of a null pointer, therefore the warning. Since this is now the only one I can reconstruct, I will drop verifier.c and keep the in-buffers (to prevent possible memory leak) in revocationinfo.c.

jhand2 commented:
@ryanhsu19 Can you reply to people's comments directly instead of making a new comment to address issues. Its too cumbersome to jump around the PR to find what you're referencing.

jhand2 commented on samples/file-encryptor/enclave/keys.cpp:
```diff
@@ -290,9 +290,17 @@ int ecall_dispatcher::prepare_encryption_header(
                 "password_key[%d] =0x%02x", i, (unsigned int)(password_key[i]));
-        goto exit;
+            goto exit;
+        }
```
Remove

ryanhsu19 replied:
Sorry I accidentally added it. I will remove it.

jhand2 commented on tests/memory/enc/basic.c:
```diff
@@ -102,11 +102,6 @@ void test_realloc(void)
-    /* realloc(X, 0) is implementation defined, but we can always free it. */
-    free(ptr);
-    ptr = realloc(NULL, 0);
-    free(ptr);
```
Why was this removed?

ryanhsu19 replied:
Reverted it. There is still a problem when attempting to release already-freed ptr at line 106 (assume ptr != NULL and ptr is released at line 100), but I will try to deal with it after other warnings are settled.

ryanhsu19 replied:
I thought about it for a while and reverted to the original version.

So there are 3 scan-build warnings, one for double-free and the others for memory leaks. Since the test is intended to ensure that memory allocation fails at the end of each unit test, it is not likely to change OE_TEST(ptr == NULL) and OE_TEST(ptr2 == NULL).

The double-free warning occurs because scan-build does not recognize oe_abort(), or it thinks that abort() is different than exit(). Because if ptr is really released at line 100, it means that ptr2 is successfully reallocated, therefore failing the test and leaving the rest of the code unreachable.

jhand2 commented on tests/syscall/cpio/commands.c:
```diff
@@ -72,7 +72,7 @@ int oe_lsr(const char* root, oe_strarr_t* paths)
     oe_strarr_release(&dirs);
 
-    if (ret != 0)
+    if (ret != 0 && paths != 0)
```
should be `paths != NULL`

ryanhsu19 replied:
Thanks. Will show up in next update

mingweishih commented on enclave/crypto/cert.c:
```diff
@@ -96,15 +96,18 @@ OE_INLINE void _referent_add_ref(Referent* referent)
 {
     /* If this was the last reference, release the object */
-    if (oe_atomic_decrement(&referent->refs) == 0)
+    if (referent)
```
For the code readability, consider do the check and early return instead of introducing nested brackets.
For example, `if (!referent) return;`

ryanhsu19 replied:
Got it. I will change it now.

ryanhsu19 commented on common/sgx/verifier.c:
```diff
@@ -118,10 +118,7 @@ static oe_result_t _add_claim(
     claim->value = (uint8_t*)oe_malloc(value_size);
     if (claim->value == NULL)
-    {
-        oe_free(claim->name);
```
I removed it because I could not reconstruct the situation and considered it unrelated. There is no scan-build error involved.

jhand2 replied:
But the oe_free line is in master. So if there is no associated warning you should leave it.

ryanhsu19 replied:
It does show "using freed memory" error in _free_claim() after I add it back, but I remember you said that oe_free(param) checks if param is null before it does anything right?

jhand2 replied:
It does, but claim->name is never set to null after the free. So before returning just do `claim->name = NULL;`

ryanhsu19 replied:
Oh, that's right. I just added it back.

jhand2 commented on samples/file-encryptor/enclave/keys.cpp:
```diff
@@ -290,9 +290,11 @@ int ecall_dispatcher::prepare_encryption_header(
             TRACE_ENCLAVE(
                 "password_key[%d] =0x%02x", i, (unsigned int)(password_key[i]));
-        goto exit;
+            goto exit;
```
Revert this back to what is in master.

ryanhsu19 replied:
Done

jhand2 commented:
Thanks! Please remove merge commits and squash commit history.

radhikaj commented:
bors r+
